Karma-Bond/Contracts/Karma-CustomBond/src/main/java/dao/karma/custombond/KarmaCustomBond.java

```diff
@@ -697,10 +700,100 @@ public BigInteger bondPrice() {
+                BigInteger newBondPrice = newTrueBondPrice.subtract(newTrueBondPrice.multiply(currentKarmaFee()).divide(MathUtils.pow10(6)));
+
+                // only apply new bond price if it is higher than the old one, this should mitigate oracle risk
+                // by defaulting to the un-capped bond price
+                if (newBondPrice.compareTo(price) > 0) {
+                    price = newBondPrice;
+                }
+            }
+        }
 
         return price;
     }
 
+    /**
+     *  Calculate bond discount using market USD prices of principal and payout token
+     *  @return BigInteger - Discount in percentages. Divide by 1e7 and multiply by 100% to convert in percentages
+     *                       on client side.
+     */
+    @External(readonly = true)
+    public BigInteger currentBondDiscount() {
+        BigInteger bondPriceUSD = this.bondPriceUSD();
+        BigInteger payoutTokenMarketPriceUSD = payoutTokenMarketPriceUSD();
+
+        // discount = (payout token market price USD - bond price USD) / payout token market price USD
+        // NOTE: result is in 1e7 decimal precision
+        return (payoutTokenMarketPriceUSD.subtract(bondPriceUSD)).divide(payoutTokenMarketPriceUSD.divide(MathUtils.pow10(7)));
+    }
+
+    /**
+     *  Calculate current bond discount
+     *  @param bondPrice - Current bond price
+     *  @return BigInteger - Discount in percentages. Divide by 1e7 and multiply by 100% to convert in percentages
+     *                       on client side.
+     */
+    @External(readonly = true)
+    public BigInteger _currentBondDiscount(BigInteger bondPrice) {
+        BigInteger bondPriceUSD = this._bondPriceUSD(bondPrice);
+        BigInteger payoutTokenMarketPriceUSD = this.payoutTokenMarketPriceUSD();
+
+        // discount = (payout token market price USD - bond price USD) / payout token market price USD
+        // NOTE: result is in 1e7 decimal precision
+        return (payoutTokenMarketPriceUSD.subtract(bondPriceUSD)).divide(payoutTokenMarketPriceUSD.divide(MathUtils.pow10(7)));
+    }
+
+    /**
+     *  Payout token market USD price pulled from Karma Oracle
+     */
+    @External(readonly = true)
+    public BigInteger payoutTokenMarketPriceUSD() {
+        // TODO init and use oracle address from constructor!
+        return IKarmaOracle.getUsdPrice(Address.fromString("cxad24e1abf6da6c401eab533433b115f476e9adf4"),
+                IToken.symbol(this.payoutToken));
+    }
+
+    /**
+     *  Calculate principal LP token market USD price, i.e. USD price of LP token
+     */
+    @External(readonly = true)
+    public BigInteger principalTokenMarketPriceUSD() {
+        // TODO init and use oracle address from constructor!
+        return IKarmaOracle.getUsdPrice(Address.fromString("cxad24e1abf6da6c401eab533433b115f476e9adf4"),
+                "USDS");

```

Comment:
Why does USDS is used for the principal token price ? Shouldn't it be `IToken.symbol(this.principalToken))` instead ?